File: tests/TestSymbolicMatrices.cpp
#include <iostream>
#include <string>
#include <cstdio>
#include <assert.h>
#include <AMD/AMD.hpp>
#include "boost/shared_ptr.hpp"

/**
 * TODO: Add more extensize tests and ensure that the results are correct.
 */ 
void testSymbolicScalarMatlab() {
  AMD::SymbolicScalarMatlab a("a");
  AMD::SymbolicScalarMatlab b("b");
  AMD::SymbolicScalarMatlab c("c");
  AMD::SymbolicScalarMatlab d = AMD::sqrt(a+b*a/c-c);
  std::string ans = std::string("sqrt((a+((b*a)/c))-c)");
  assert( d.getString() == ans );
  AMD::SymbolicScalarMatlab e(1.2);
  ans = std::string("1.2");
  assert(e.getString() == ans);
}

/**
 * TODO: Add more extensive tests and ensure that the results are correct.
 */ 
void testSymbolicMatrixMatlab() {
  AMD::SymbolicMatrixMatlab a("A");
  AMD::SymbolicMatrixMatlab b("B");
  AMD::SymbolicMatrixMatlab c("C");
  AMD::SymbolicMatrixMatlab d = AMD::inv(a+b*(AMD::transpose(a)-c));
  std::string ans = std::string("inv(A+(B*(A'-C)))");
  assert(d.getString() == ans);

  AMD::SymbolicScalarMatlab e = AMD::trace(AMD::transpose(a+b));
  ans = "trace((A+B)')";
  assert(e.getString() == ans);

  e = AMD::logdet(a-b*c);
  ans = "log(det(A-(B*C)))";
  assert(e.getString() == ans);

  e = AMD::fnorm(a+AMD::elementwiseProd(b,c));
  ans = "norm(A+(B.*C),'fro')";
  assert( e.getString() == ans );

  AMD::SymbolicScalarMatlab f("f");
  b = f*a+a/f;
  ans = "((f.*A)+(A./f))";
  assert(b.getString() == ans);

  b = a*f;
  ans = "(f.*A)";
  assert(b.getString() == ans);

  // check sizes
  AMD::SymbolicMatrixMatlab x("X",2,3);
  AMD::SymbolicMatrixMatlab y("B",3,5);
  AMD::SymbolicMatrixMatlab z("C",2,5);
  AMD::SymbolicMatrixMatlab u = transpose(x*y+z);
  assert(u.getNumRows()==5 && u.getNumCols()==2);
}


int main() {
  std::cout << "Testing SymbolicScalarMatlab .... ";
  testSymbolicScalarMatlab();
  std::cout << "DONE" << std::endl;

  std::cout << "Testing SymbolicScalarMatlab .... ";
  testSymbolicMatrixMatlab();
  std::cout << "DONE" << std::endl;

  std::cout << "All tests passed." << std::endl;

  return(0);
}

File: tests/python/test_symbolic_matrices.py
" Test Symbolic Matrices Python binding. "

from setup import *

def test_SymScal():
    """ Test SymScal (SymbolicScalarMatlab) """
    print 'Testing SymScal...'
    a = SymScal('a')
    b = SymScal('b')
    c = SymScal('c')
    d = sqrt(a + b * a / c - c)
    assert d.string == "sqrt((a+((b*a)/c))-c)"

    e = SymScal(1.2)
    assert e.string == '1.2'

    print 'a = ', a
    print 'b = ', b
    print 'c = ', c
    print 'd = ', d

def test_SymMat():
    """ Test SymMat (SymbolicMatrixMatlab) """
    print 'Testing SymMat...'
    a = SymMat("A")
    b = SymMat("B")
    c = SymMat("C")
    d = inv(a + b * (transpose(a) - c))
    assert d.string == "inv(A+(B*(A'-C)))"

    print 'A = ', a
    print 'B = ', b
    print 'C = ', c
    print 'D = ', d
    
    e = trace(transpose(a + b))
    assert e.string == "trace((A+B)')"

    e = logdet(a - b * c)
    assert e.string == "log(det(A-(B*C)))"

    e = fnorm(a + elementwise_product(b, c))
    assert e.string == "norm(A+(B.*C),'fro')"

    f = SymScal("f")
    b = f * a + a / f
    assert b.string == "((f.*A)+(A./f))"

    b = a * f
    assert b.string == "(f.*A)"

    #check sizes
    x = SymMat("X", 2, 3)
    y = SymMat("B", 3, 5)
    z = SymMat("C", 2, 5)
    u = transpose(x * y + z)
    assert u.num_rows == 5 and u.num_cols == 2

    

def main():
    """ Test Symbolic Matrices Python binding """
    test_SymScal()
    test_SymMat()
    print 'Done.'

if __name__ == '__main__':
    main()

